### PR TITLE
chore: Support multiple render backends

### PR DESCRIPTION
**Pull request by fredizzimo**
## What kind of change does this PR introduce?
This is mostly a refactoring in order to support multiple render backends, like D3D, and makes the actual PR that implements the D3D backend much smaller.

* WindowContext has been renamed to `SkiaRenderer`, which can have different implementations depending on the platform
* The Gpu profiling system now supports multiple backends, and code has been moved to a dedicated `opengl` implementation
* `VSyncWin` has been renamed to `VSyncWinDwm` to allow a swapchain implementation for D3D.
* The latency reduction optimization has been removed, as it is incompatible with some render backends.

## Did this PR introduce a breaking change? 
- No

### CODE REVIEW

File: src/renderer/opengl.rs [hunk 1]
```diff
@@ -16,63 +17,202 @@ use glutin::{
...
 
-pub struct Context {
-    surface: Surface<WindowSurface>,
+use crate::{profiling::tracy_gpu_zone, window::UserEvent};
+
+#[cfg(feature = "gpu_profiling")]
+use crate::profiling::{opengl::create_opengl_gpu_context, GpuCtx};
+
+pub struct OpenGLSkiaRenderer {
+    // NOTE: The destruction order is important, so don't re-arrange
```
Comment (MultisampledNight) [inline src/renderer/opengl.rs:48]: Maybe a `SAFETY` could fit better in order to convey that a critical invariant is upheld here.

File: src/renderer/opengl.rs [hunk 2]
```diff
@@ -16,63 +17,202 @@ use glutin::{
...
+        let raw_window_handle = window.raw_window_handle();
+
+        let size = clamp_render_buffer_size(&window.inner_size());
+
+        let surface_attributes = SurfaceAttributesBuilder::<WindowSurface>::new()
+            .with_srgb(Some(srgb))
+            .build(
+                raw_window_handle,
+                NonZeroU32::new(size.width).unwrap(),
+                NonZeroU32::new(size.height).unwrap(),
```
Comment (MultisampledNight) [inline src/renderer/opengl.rs:89]: Those unwraps might very well blow up if starting in e.g. an initially minimized state, need to try that out.

File: src/renderer/opengl.rs [hunk 3]
```diff
@@ -16,63 +17,202 @@ use glutin::{
...
+            window_surface.set_swap_interval(&context, SwapInterval::DontWait)
+        };
 
-    pub fn resize(&self, width: NonZeroU32, height: NonZeroU32) {
-        GlSurface::resize(&self.surface, &self.context, width, height)
+        gl::load_with(|s| get_proc_address(&window_surface, CString::new(s).unwrap().as_c_str()));
+
+        let interface = skia_safe::gpu::gl::Interface::new_load_with(|name| {
+            if name == "eglGetCurrentDisplay" {
+                return std::ptr::null();
```
Comment (MultisampledNight) [inline src/renderer/opengl.rs:118]: Why is `eglGetCurrentDisplay` filtered out? Can't we let `get_proc_address` also find `eglGetCurrentDisplay`? (possibly this is a peculiarity of EGL, I've never used EGL)

**Review by MultisampledNight — COMMENTED** [review]
Nothing too interesting code-wise, typical noisy GL setup. Haven't tested yet.

File: src/renderer/vsync/mod.rs [hunk 4]
```diff
@@ -4,59 +4,43 @@ mod macos_display_link;
...
 
 use crate::{
-    renderer::WindowedContext, settings::SETTINGS, window::UserEvent, window::WindowSettings,
+    renderer::SkiaRenderer, settings::SETTINGS, window::UserEvent, window::WindowSettings,
 };
-use winit::event_loop::EventLoopProxy;
+use winit::{event_loop::EventLoopProxy, window::Window};
 
 #[cfg(target_os = "linux")]
 use std::env;
```
Comment (MultisampledNight) [inline src/renderer/vsync/mod.rs:17]: Fwiw due to the moved detection logic, this import is actually unused and yields a warning

```rust
warning: unused import: `std::env`
  --> src/renderer/vsync/mod.rs:17:5
   |
17 | use std::env;
   |     ^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
```

**Comment by MultisampledNight** [comment]
Test results:

- On native wayland, everything's fine and smooth as ever.
- On Xwayland, there's a few freezes the first few seconds after starting up, where interestingly the view keeps updating (the profiler keeps moving and displaying), but the entire text contents are not responding to any keypresses for a few seconds. Maybe this doesn't come from this PR though, I haven't tested main in a while.

**Comment by MultisampledNight** [comment]
After retrying a few times between main and this branch on Xwayland, I can't seem to reproduce what I found initially, and found no meaningful differences. Ignore that then.

File: src/renderer/opengl.rs (hunk 3 above)
Comment (fredizzimo) [reply src/renderer/opengl.rs:118]: I have no idea, it was introduced here 
* https://github.com/neovide/neovide/pull/631

After some discussion here
* https://github.com/neovide/neovide/issues/491#issuecomment-830762584
* https://github.com/neovide/neovide/issues/661

I found no issues in the glutin repository.

File: src/renderer/opengl.rs (hunk 2 above)
Comment (fredizzimo) [reply src/renderer/opengl.rs:89]: The code is just moved, so it has been the same before. `clamp_render_buffer_size` should ensure that it's not zero.

**Comment by fredizzimo** [comment]
I'm not sure about the XWayland freeze, I can't think of any changes that could cause it, except this one https://github.com/neovide/neovide/pull/2359/commits/110ea9957a003f1fe85eec53542be0e19474da97, but that should not affect x11 or xwayland at all. 

The rest of the code should have stayed functionally the same.

File: src/renderer/opengl.rs (hunk 1 above)
Comment (fredizzimo) [reply src/renderer/opengl.rs:48]: I'll keep the `NOTE:` the syntax highlighter highligts it, but not `SAFETY:`

File: src/renderer/vsync/mod.rs (hunk 4 above)
Comment (fredizzimo) [reply src/renderer/vsync/mod.rs]: That is now fixed. I was doing most of my testing on Windows, and that was part of one late change that never got properly tested on Linux.

**Pull request merged**